tsdoc/src/beta/DeclarationReference.ts

```diff
@@ -55,8 +55,15 @@ export class DeclarationReference {
 
   public static makeSafeComponent(text: string): string {
```

Comment:
@rbuckton in https://github.com/microsoft/web-build-tools/pull/1406#discussion_r305667609 I am asking that `.addNavigationStep()` should treat its input as an unquoted identifier, rather than parsing it.

Current design:  
- `.addNavigationStep("[X]")`  --> `[X]`  (i.e. an ECMAScript symbol)
- `.addNavigationStep("[X")`  --> `"[X"`  (i.e. a quoted identifier)
- `.addNavigationStep("\"X\"")`  --> `"X"`  
- `.addNavigationStep("X")`  --> `X`  

What I think is a better design:
- `.addNavigationStep("[X]")`  --> `"[X]"`  (still an identifier, not reinterpreted as an ECMAScript symbol expression)
- `.addNavigationStep("[X")`  --> `"[X"`
- `.addNavigationStep("\"X\"")`  --> `X`  (quotes removed because they are redundant)
- `.addNavigationStep("X")`  --> `X`

Reply:
My proposal should also eliminate the unexceptional exception (`catch { // do nothing }`).

We can probably use [StringChecks.explainIfInvalidUnquotedIdentifier()](https://github.com/microsoft/tsdoc/blob/858cbc136969ecabb0c236ab5b0cfe6a3af2daf7/tsdoc/src/parser/StringChecks.ts#L154) to determine whether quotes are needed.

Reply:
> - `.addNavigationStep("[X]") --> `"[X]"` 

I completely disagree. One of the things I wanted to fix with all of this was having proper names for computed property names, as I have a large number of APIs that I am using api-extractor/api-documenter for that use `unique symbol` types, and I fully intended for this to address those use cases as well.

> - `.addNavigationStep("\"X\"") --> `X`

I also disagree with this. It should be what was written, or encoded if what was written isn't a valid component. 

Consider a URL: If I create a URL from `"http://foo.com/bar%2520baz"` and decode it, then I end up with `"http://foo.com/bar%20baz"`, which could easily be misinterpreted as `"http://foo.com/bar baz"` if you pass it to another API. However, if I create a URL from `"http://foo.com/bar baz"` it *must* be encoded as `"http://foo.com/bar%20baz"` because the space character is not permitted.

I think we should reserve quotes by default, so a quoted string containing a quoted string should be escaped:

- `.addNavigationStep("[X]")` -> `[X]`
- `.addNavigationStep("[X")` -> `"[X"`
- `.addNavigationStep("\"X\"")` -> `"\"X\""`
- `.addNavigationStep("X") -> `X`

If we need to pass through quotes, we should add an optional argument to `addNavigationStep` to indicate the string is already escaped (in which case we should only validate it).

Reply:
Ahh...  I did not realize that in your `DeclarationReference` API, `.addNavigationStep("[X]")` is the only way to make an ECMAScript symbol.  I was confused because in the TSDoc parser, the AST and API have special nodes for representing an ECMAScript symbol and its internal components.

For example, this:

```ts
/** {@link my-control-library#Button.[UISymbols.toNumberPrimitive]} */
```

...gets parsed as this (in the [TSDoc playground](https://microsoft.github.io/tsdoc/#)):
```
- LinkTag
  * InlineTag_OpeningDelimiter="{"
  * InlineTag_TagName="@link"
  * Spacing=" "
  - DeclarationReference
    * DeclarationReference_PackageName="my-control-library"
    * DeclarationReference_ImportHash="#"
    - MemberReference
      - MemberIdentifier
        * MemberIdentifier_Identifier="Button"
    - MemberReference
      * MemberReference_Dot="."
      - MemberSymbol
        * DocMemberSymbol_LeftBracket="["
        - DeclarationReference
          - MemberReference
            - MemberIdentifier
              * MemberIdentifier_Identifier="UISymbols"
          - MemberReference
            * MemberReference_Dot="."
            - MemberIdentifier
              * MemberIdentifier_Identifier="toNumberPrimitive"
        * DocMemberSymbol_RightBracket="]"
  * InlineTag_ClosingDelimiter="}"
```

The `UISymbols.toNumberPrimitive` substring becomes a second declaration reference nested inside the outer one, and that is what TSDoc's [DocMemberSymbol constructor](https://github.com/microsoft/tsdoc/blob/858cbc136969ecabb0c236ab5b0cfe6a3af2daf7/tsdoc/src/nodes/DocMemberSymbol.ts#L10) API receives.

For a "beta" prototype, I'm fine with your current approach of representing a symbol as a string, but we would need to fix it longer term.  Otherwise it may be awkward for the caller to construct an input like `[a.[b]."[c]"]`.  (The TypeScript compiler may provide utilities that help build expressions like this, but strictly speaking, it's making TypeScript expressions not UID expressions; there may be edge cases where the escaping rules differ.  Ideally the TypeScript AST need to be transformed into TSDoc AST.)

> I completely disagree. One of the things I wanted to fix with all of this was having proper names for computed property names, as I have a large number of APIs that I am using api-extractor/api-documenter for that use `unique symbol` types, and I fully intended for this to address those use cases as well.

I agree that ECMAScript symbols should be handled better by API Extractor.  But the current implementation is somewhat of a hack.  Here's how the `name` .api.json field ends up for some various inputs:

```ts
export declare const c: unique symbol;
export declare class X {
    "a": string;    // "name": "a"
    "b[": string;   // "name": "b["
    [c]: string;    // "name": "[c]"
    d: string;      // "name": "d"
    "[e]": string;  // "name": "[e]"
}
```

The reason is that [this code](https://github.com/microsoft/web-build-tools/blob/c5e13a01bf88d5c08def2db5b46f27df97083ac7/apps/api-extractor/src/analyzer/AstSymbolTable.ts#L447) calls `tryGetLateBoundName()` only when a symbol is encountered, which is technically mixing escaping contexts.

If we pass this through to `addNavigationStep()` directly then `"[e]"` will get misinterpreted.

> I think we should reserve quotes by default, so a quoted string containing a quoted string should be escaped:
> 
> * `.addNavigationStep("[X]")` -> `[X]`
> * `.addNavigationStep("[X")` -> `"[X"`
> * `.addNavigationStep("\"X\"")` -> `"\"X\""`
> * `.addNavigationStep("X")` -> `X`
> 
> If we need to pass through quotes, we should add an optional argument to `addNavigationStep` to indicate the string is already escaped (in which case we should only validate it).

If we're saying that `addNavigationStep()` receives a TypeScript-like expression, then `.addNavigationStep("[X")` should throw an exception.  It is not a valid expression.  Silently adding quotes to it will hide an escaping mistake in the caller, which is likely to cause even more confusion.  In my example above, API Extractor is escaping incorrectly.  We need to fix that, not patch it up in one particular code path.

Reply:
TypeScript only allows late-bound symbols that can be referenced via an entity name, so only `a` or `a.b.c` are allowed. Element access is disallowed in an entity name, so `a.[b]."[c]"` wouldn't be supported.

> If we pass this through to addNavigationStep() directly then "[e]" will get misinterpreted.

If that's true, then we should also always quote anything with `[]` and require the user to provide a value to an optional argument indicating the user did the escaping work. I will have something for that in an update shortly.

Reply:
If we're going to represent ECMAScript symbol expressions as strings, then it seems that API Extractor needs to emit the `name` field like this:

```ts
export declare class X {
    "a": string;    // "name": "a"        <--- remove redundant quotes
    "b[": string;   // "name": "\"b[\"" <--- but quote identifiers with bad characters
    [c]: string;    // "name": "[c]" 
    d: string;      // "name": "d"
    "[e]": string;  // "name": "\"[e]\""
}
```
And these would be the inputs for `addNavigationStep()`.

Reply:
`addNavigationStep` and a few other methods now have a `userEscaped` flag. The default is false, in which case everything is passed through `DeclarationReference.escapeComponentString` (which adds quotes around any string with reserved characters including `"` and strings starting with `[`).

If a user passes `true` to `userEscaped`, we merely validate that the string is a valid component and perform no escaping.

Reply:
We could possibly consider allowing more context to the content inside of brackets (such as a source prefix like `package!`), though disambiguation probably isn't necessary (it has to be either a `var` or a `property` to be valid, and those two meanings cannot collide).

Reply:
Perhaps it should just be `` `[` DeclarationReference `]` `` and any other interpretation must be quoted.

Reply:
> TypeScript only allows late-bound symbols that can be referenced via an entity name, so only `a` or `a.b.c` are allowed. Element access is disallowed in an entity name, so `a.[b]."[c]"` wouldn't be supported.

This is an wonderful (and uncharacteristic of JavaScript) constraint to impose.  Even `public static readonly "[f]": unique symbol = Symbol("d");` is not allowed.  It really simplifies things!

Reply:
Hrmm... but consider this example:

Entry point for **@<!-- -->scope/my-npm-package**
```ts
export namespace MySymbols {
  export const MySymbol: unique symbol = Symbol("MySymbol");
}
```

Entry point for **this-package**:
```ts
import { MySymbols } from '@scope/my-npm-package';

export class X {
  // this-package!X.[@scope/my-npm-package!MySymbols.MySymbol]:member
  [MySymbols.MySymbol]: string;
}
```
Even though TypeScript requires `[MySymbols.MySymbol]` to be simple, the declaration reference normally does not refer to imported symbols.  Thus it needs a qualifier to explain which `MySymbols` we're referring to.   :-/

Reply:
We don't disallow `public static readonly "x-y": unique symbol`, but we do disallow it as a reference in a computed property name:

```ts
interface Foo {
  [MySymbols["x-y"]]: number; // this is an error.
}
```

Reply:
Seems like maybe the right long term solution really is for .api.json to replace the `name` field with a nesting expression, and for your API to provide a way to build up these expressions and visit their components.

Today API Extractor stupidly treats symbols as quoted text strings.  In other words:

`this-package!X#"[MySymbols.MySymbol]":member`

This is incorrect, but from a documentation standpoint, it seems to produce DocFX-compatible output that looks decent to humans, for 90% of the cases that arise in a typical project.  Perhaps we should maintain this _status quo_ so we can get your feature rolled out.  Then come back and tackle ECMAScript symbols as an advanced feature.

Reply:
> ```ts
> import { MySymbols } from '@scope/my-npm-package';
> 
> export class X {
>   // this-package!X#[@scope/my-npm-package!MySymbols.MySymbol]:member
>   [MySymbols.MySymbol]: string;
> }
> ```
> ... Thus it needs a qualifier to explain which MySymbols we're referring to. :-/

I'm unclear as to whether you mean the ref above is a good thing or a bad thing.

If I change the *BracketedComponent* production to `` `[` DeclarationReference `]` `` then it would be able to parse the above ref. Would that be preferred?

Reply:
Yes. From the perspective of the grammar, I think `this-package!X#[@scope/my-npm-package!MySymbols.MySymbol]:member` is how we need to represent these things.

But I'm also realizing that's nontrivial work for API Extractor to build them.  The `this-package!X#"[MySymbols.MySymbol]":member` hack is safe and reliable.  So I think we need to procrastinate nesting references, i.e. always set `userEscaped=true` for now.

Reply:
I suppose `this-package!X#[MySymbols.MySymbol]:member` could also work as a hack.  Maybe that's what you were suggesting.

Reply:
I have an update that does `` `[` DeclarationReference `]` `` that I can push up. For computed property names and builtin symbols, the difference would be something like this:

```ts
// symbolName = "[Symbol.iterator]";
return DeclarationReference
  .from(this.parent && this.parent.canonicalReference)
  .addNavigationStep(Navigation.Exports, DeclarationReference.parse(symbolName))
  .withMeaning(Meaning.Member);
```

(I added `DeclarationReference.from` as a convenience method, it returns `.empty()` if the argument is `undefined`).

Reply:
Also, `X#"[MemberSymbols.MySymbol]"` isn't safe, since someone could *theoretically* create a string-valued property with that name as well. It's highly unlikely, but possible.

Reply:
I pushed an update to PR https://github.com/microsoft/web-build-tools/pull/1406 showing how the output looks with the new selectors.

I'm investigating how hard it is to get API Extractor to make the names like this:
```ts
export declare class X {
    "a": string;    // "name": "a"
    "b[": string;   // "name": "\"b[\""
    [c]: string;    // "name": "[c]"
    d: string;      // "name": "d"
    "[e]": string;  // "name": "\"[e]\""
}
```

It may need to be a separate PR, since there seem to be a number of places where names are built (and currently only 2 places consider symbols).

Reply:
Heh quoting can be declaration-specific:
```ts
class X {
  public "f1"(x: string): void;
  public f1(x: boolean): void;
  public f1(x: string | boolean): void {
  }
}
```

Reply:
This seems to work okay:

```ts
if (StringChecks.explainIfInvalidUnquotedIdentifier(followedSymbol.name)) {
  localName = JSON.stringify(followedSymbol.name);  //<====
} else {
  localName = followedSymbol.name;
}

if (TypeScriptHelpers.isWellKnownSymbolName(followedSymbol.name)) {
  // TypeScript binds well-known ECMAScript symbols like "Symbol.iterator" as "__@iterator".
  // This converts a string like "__@iterator" into the property name "[Symbol.iterator]".
  localName = `[Symbol.${localName.slice(3)}]`;
} else {
  const isUniqueSymbol: boolean = TypeScriptHelpers.isUniqueSymbolName(followedSymbol.name);
  for (const declaration of followedSymbol.declarations || []) {
    const declarationName: ts.DeclarationName | undefined = ts.getNameOfDeclaration(declaration);
    if (declarationName && ts.isIdentifier(declarationName)) {
      localName = declarationName.getText().trim();
      break;
    }
    if (isUniqueSymbol && declarationName && ts.isComputedPropertyName(declarationName)) {
      const lateBoundName: string | undefined = TypeScriptHelpers.tryGetLateBoundName(declarationName);
      if (lateBoundName) {
        localName = lateBoundName;
        break;
      }
    }
  }
}
```

If you don't see any problems with it, I'll open a separate PR that moves this name-calculator into utility, and fixes up everywhere that names get made.

Reply:
Seems like symbols and malformed identifiers can only occur as (possibly `static`) members of a `class` or `interface`.  Thus there are less places to worry about.

Reply:
Here's the PR: https://github.com/microsoft/web-build-tools/pull/1410